src/libsnark.rs

```diff
@@ -51,21 +54,68 @@ pub fn setup<T: Field> (
     let num_constraints = a.len();
     let num_variables = variables.len();
 
-    //initialize matrix entries with 0s.
-    let mut a_arr: Vec<[u8; 32]> = vec![[0u8; 32]; num_constraints * num_variables];
-    let mut b_arr: Vec<[u8; 32]> = vec![[0u8; 32]; num_constraints * num_variables];
-    let mut c_arr: Vec<[u8; 32]> = vec![[0u8; 32]; num_constraints * num_variables];
-
+    // Create single A,B,C vectors of tuples (constraint_number, variable_id, variable_value)
+    let mut a_vec = vec![];
+    let mut b_vec = vec![];
+    let mut c_vec = vec![];
     for row in 0..num_constraints {
-        for &(idx, ref val) in &a[row] {
-            a_arr[row * num_variables + idx] = vec_as_u8_32_array(&val.into_byte_vector());
-        }
-        for &(idx, ref val) in &b[row] {
-            b_arr[row * num_variables + idx] = vec_as_u8_32_array(&val.into_byte_vector());
-        }
-        for &(idx, ref val) in &c[row] {
-            c_arr[row * num_variables + idx] = vec_as_u8_32_array(&val.into_byte_vector());
-        }
+      for &(idx, ref val) in &a[row] {
+          a_vec.push((row as i32, idx as i32, vec_as_u8_32_array(&val.into_byte_vector())));
+      }
+      for &(idx, ref val) in &b[row] {
+          b_vec.push((row as i32, idx as i32, vec_as_u8_32_array(&val.into_byte_vector())));
+      }
+      for &(idx, ref val) in &c[row] {
+          c_vec.push((row as i32, idx as i32, vec_as_u8_32_array(&val.into_byte_vector())));
+      }
+    }
+
+    // Convert above A,B,C vectors to byte arrays for cpp
+    let mut a_arr: Vec<u8> = vec![0u8; 40 * a_vec.len()];
```

Comment:
@lubieplacki could you add a comment or define variables for 40, 4, 8?

Reply:
:+1: